### PR TITLE
Supg incompressible fluid solver

### PR DESCRIPTION
**Pull request by icbannerman**

### CODE REVIEW

File: source/mpi_insim_supg.cpp [hunk 1]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
     template <int dim>
     void SUPGInsIM<dim>::assemble(const bool use_nonzero_constraints)
     {
+        TimerOutput::Scope timer_section(timer, "Assemble system");
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:17]: format

File: source/mpi_insim_supg.cpp [hunk 2]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
...
+      std::vector<Tensor<1, dim>> fsi_acc_values(n_q_points);
+
+      std::vector<double> div_phi_u(dofs_per_cell);
+      std::vector<Tensor<1, dim>> phi_u(dofs_per_cell);
+      std::vector<Tensor<2, dim>> grad_phi_u(dofs_per_cell);
+      std::vector<double> phi_p(dofs_per_cell);
+      std::vector<Tensor<1, dim>> grad_phi_p(dofs_per_cell);
+
+      // Zero out sigma field and body force if their fields are not specified
+      if (sigma_pml_field == nullptr)
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:80]: Remove this pml stuff, not used

File: source/mpi_insim_supg.cpp [hunk 3]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
...
+              fe_values[pressure].get_function_gradients(
+                evaluation_point, current_pressure_gradients);
+
+              fe_values[velocities].get_function_values(
+                present_solution, present_velocity_values);
+
+              fe_values[pressure].get_function_values(present_solution,
+                                                      present_pressure_values);
+
+              if (sigma_pml_field)
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:119]: remove

File: source/mpi_insim_supg.cpp [hunk 4]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
...
+      // are
+      // all obtained through their shape functions at quadrature points.
+
+      std::vector<Tensor<1, dim>> current_velocity_values(n_q_points);
+      std::vector<Tensor<2, dim>> current_velocity_gradients(n_q_points);
+      std::vector<double> current_pressure_values(n_q_points);
+      std::vector<Tensor<1, dim>> current_pressure_gradients(n_q_points);
+      std::vector<Tensor<1, dim>> present_velocity_values(n_q_points);
+      std::vector<double> present_pressure_values(n_q_points);
+      std::vector<double> sigma_pml(n_q_points);
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:69]: Remove

File: source/mpi_insim_supg.cpp [hunk 5]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
...
+      std::vector<Tensor<1, dim>> artificial_bf(n_q_points);
+      std::vector<Tensor<1, dim>> fsi_acc_values(n_q_points);
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:71]: For 63-71: I'm certain some of the variables you declare here are not used. Check and remove them.

File: source/mpi_insim_supg.cpp [hunk 6]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
...
+              fe_values[pressure].get_function_values(evaluation_point,
+                                                      current_pressure_values);
+
+              fe_values[pressure].get_function_gradients(
+                evaluation_point, current_pressure_gradients);
+
+              fe_values[velocities].get_function_values(
+                present_solution, present_velocity_values);
+
+              fe_values[pressure].get_function_values(present_solution,
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:116]: 101-116: Same here. remove unused variables

File: source/mpi_insim_supg.cpp [hunk 7]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
...
+                  body_force->tensor_value_list(
+                    fe_values.get_quadrature_points(), artificial_bf);
+                }
+
+              fe_values[velocities].get_function_values(fsi_acceleration,
+                                                        fsi_acc_values);
+
+              for (unsigned int q = 0; q < n_q_points; ++q)
+                {
+                  const double rho = parameters.fluid_rho *
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:135]: remove ``ind``

File: source/mpi_insim_supg.cpp [hunk 8]
```diff
@@ -14,6 +14,362 @@ namespace Fluid
...
+                            current_velocity_values[q] * phi_u[i] +
+                          current_pressure_values[q] * div_phi_u[i]) -
+                         rho *
+                           (current_velocity_values[q] -
+                            present_velocity_values[q]) *
+                           phi_u[i] / time.get_delta_t() +
+                         (gravity + artificial_bf[q]) * phi_u[i] * rho) *
+                        fe_values.JxW(q);
+                      local_rhs(i) +=
+                            -(current_velocity_divergence * phi_p[i] )* fe_values.JxW(q);
```
Comment (yufeimi) [inline source/mpi_insim_supg.cpp:283]: format!

File: tests/couette_flow_mpi_insim_supg/couette_flow_mpi_insim_supg.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+#include "parameters.h"
+#include "utilities.h"
+
+extern template class Fluid::MPI::SUPGInsIM<2>;
+extern template class Fluid::MPI::SUPGInsIM<3>;
+
+using namespace dealii;
+
+int main(int argc, char *argv[])
+
```
Comment (yufeimi) [inline tests/couette_flow_mpi_insim_supg/couette_flow_mpi_insim_supg.cpp:26]: remove nonsense blank lines. Once again, format

File: tests/couette_flow_mpi_insim_supg/couette_flow_mpi_insim_supg.cpp [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+      if (argc > 1)
+        {
+          infile = argv[1];
+        }
+      Parameters::AllParameters params(infile);
+
+      double L = 2, D = 0.4, d = 0.025, l = 0.1;
+      if (params.dimension == 2)
+        {
+
```
Comment (yufeimi) [inline tests/couette_flow_mpi_insim_supg/couette_flow_mpi_insim_supg.cpp:44]: nonsense blank line

**Review by yufeimi — CHANGES_REQUESTED** [review]
1. Too many commits. Merge them. No need for one commit for each test.
2. Your test doesn't check anything about the solution, which makes the tests meaningless. They are simple cases and are almost certain to run.
3. Format.

File: tests/couette_flow_mpi_insim_supg/couette_flow_mpi_insim_supg.cpp [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+            {static_cast<unsigned int>(L / l),
+             static_cast<unsigned int>(D / d)},
+            Point<2>(0, 0),
+            Point<2>(L, D),
+            true);
+          Fluid::MPI::SUPGInsIM<2> flow(tria, params);
+          flow.run();
+          auto solution = flow.get_current_solution();
+        }
+
```
[the remark this answers is not in the document]
Comment (icbannerman) [reply tests/couette_flow_mpi_insim_supg/couette_flow_mpi_insim_supg.cpp:57]: thanks for the letting me know all these errors ; I will correct them

**Review by yufeimi — CHANGES_REQUESTED** [review]

File: tests/fluid_plane_wall_driven_mpi_insim_supg/fluid_plane_wall_driven_mpi_insim_supg.prm [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+  # List the constrained components of these boundaries
+  # One decimal number indicates one set of constrained components:
+  # 1-x, 2-y, 3-xy, 4-z, 5-xz, 6-yz, 7-xyz
+  # To make sense of the numbering, convert decimals to binaries (zyx)
+  set Dirichlet boundary components = 3, 3
+
+  # Specify the values of the Dirichlet BCs, including both homogeneous and
+  # inhomogeneous ones.
+  set Dirichlet boundary values = 0, 0, 1 , 0
```
Comment (yufeimi) [inline tests/fluid_plane_wall_driven_mpi_insim_supg/fluid_plane_wall_driven_mpi_insim_supg.prm:89]: no space before comma

File: tests/fluid_plane_wall_driven_mpi_insim_supg/fluid_plane_wall_driven_mpi_insim_supg.cpp [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+          Fluid::MPI::SUPGInsIM<2> flow(tria, params);
+          flow.run();
+          auto solution = flow.get_current_solution();
+          // Check the max values of velocity . The final velocity profile is a
+          // striaght line with positive gradient.
+          // The maximum velocity is equal to the upper plate boundary
+          // condition.
+          auto v = solution.block(0);
+          double vmax = Utils::PETScVectorMax(v);
+          double verror = std::abs(vmax - 1.0);
```
Comment (yufeimi) [inline tests/fluid_plane_wall_driven_mpi_insim_supg/fluid_plane_wall_driven_mpi_insim_supg.cpp:51]: This is a never-failing test, so it's meaningless. Think about why.

File: tests/fluid_plane_wall_driven_mpi_insim_supg/fluid_plane_wall_driven_mpi_insim_supg.cpp [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            tria,
+            {static_cast<unsigned int>(L / l),
+             static_cast<unsigned int>(D / d)},
+            Point<2>(0, 0),
+            Point<2>(L, D),
+            true);
+          Fluid::MPI::SUPGInsIM<2> flow(tria, params);
+          flow.run();
+          auto solution = flow.get_current_solution();
+          // Check the max values of velocity . The final velocity profile is a
```
Comment (yufeimi) [inline tests/fluid_plane_wall_driven_mpi_insim_supg/fluid_plane_wall_driven_mpi_insim_supg.cpp:45]: no space before period

**Review by yufeimi — CHANGES_REQUESTED** [review]
Check the sentences you wrote. No space before comma or period.
